src/core/lib/surface/call.cc

```diff
@@ -1030,6 +1030,12 @@ static void post_batch_completion(batch_control* bctl) {
   if (bctl->op.send_trailing_metadata) {
     call->send_trailing_metadata.Clear();
   }
+  if (bctl->op.recv_initial_metadata && !bctl->op.recv_trailing_metadata) {
+    if (call->is_client && call->is_trailers_only) {
+      error = grpc_error_add_child(error, GRPC_ERROR_CREATE_FROM_STATIC_STRING(
```

Comment:
If we have the transport return an error then that would result in the surface starting a cancel op down the stack which could overwrite the original fake status. For example, if the call original failed with deadline exceeded, it would now fail with a cancelled status

> Also, if we're going to return an error for this op back to through the surface API, then that implies that we can back out the changes in #26379.
I had considered this and I was going to remove it but then I considered the case where the batch that includes the initial metadata op also includes the recv status op, and in that case we set the error to GRPC_ERROR_NONE, so the upper layer won't know whether initial metadata was received or not.

Reply:
> If we have the transport return an error then that would result in the surface starting a cancel op down the stack which could overwrite the original fake status. For example, if the call original failed with deadline exceeded, it would now fail with a cancelled status

Shouldn't the transport use whichever status happened first?

> I had considered this and I was going to remove it but then I considered the case where the batch that includes the initial metadata op also includes the recv status op, and in that case we set the error to GRPC_ERROR_NONE, so the upper layer won't know whether initial metadata was received or not.. which seems to be relevant information for the C++ callback API.

Does the callback API ever put the recv_initial_metadata and recv_trailing_metadata ops in the same batch?

Reply:
> Shouldn't the transport use whichever status happened first?

We could maybe, but we've got super old code that advocates the other way around https://github.com/grpc/grpc/blob/87068e02c3c4d34f51019b3ccdceea1230c92fea/src/core/ext/transport/chttp2/transport/chttp2_transport.cc#L2043 @ctiller might know the context here better. Also, I'm not sure if how that would with the new promise API

Reply:
> Does the callback API ever put the recv_initial_metadata and recv_trailing_metadata ops in the same batch?

It does
https://github.com/grpc/grpc/blob/87068e02c3c4d34f51019b3ccdceea1230c92fea/include/grpcpp/impl/codegen/client_callback.h#L76
Hmm, but this instance does not care for the success status of the op, so maybe it's ok to revert that change.

Reply:
> Also, I'm not sure if how that would with the new promise API

Craig can jump in here, but from previous conversations with him, I don't think this will be a problem in the promise world, because there should no longer be a need for the surface to send down a cancel op when one of the batches fails; instead, cancellation will work by simply destroying the entire promise tree, at which point there will no longer be any opportunity for the call to even see the status returned by the transport (i.e., the cancellation status will always take precedence).  So I think the entire question of the transport deciding precedence will likely go away; whatever failure happens first will determine the status of the call.

I suspect it would be fine to go ahead and change the transport to have this same first-failure-wins behavior now.  But again, I'll let Craig weigh in on that.

Reply:
Captain context here!

So Mark's right on the promise world: if we cancel somewhere, we simply destroy the lower part of the tree and choose a status at that point.

This tends to be the most natural way to define error handling (higher layer things get more say in the decision) and is the reason for the at-first-odd looking rules in transport. Since we have to push errors down to the bottom of the stack to properly publish them (ugh) we need to choose a rule down at the bottom for which error wins that mimics what our top level intuition is.

The temporally first error observed by the transport is not that: suppose the transport gets busy and then observes an write error, while concurrently the application decides to cancel the request due to an authentication problem - it's more useful for us to report the authentication problem to the application - so that should dominate.

Hence the rule in the transport today: always prefer the structurally highest level error. This will be the behavior in the new promises world too (indeed its going to be terrifically hard to write anything else in that world).

Reply:
Thanks, Craig, that's useful.  Good point about us wanting the status from the highest point in the stack, not the one seen by the transport first.

I'm now wondering if we should just defer this PR until after the promise conversion.  At that point, I think it will be much easier to make this change in a clean way.  We'll probably be able to just remove the `trailing_metadata_available` field completely from the transport API.

Reply:
Given that we've identified broken API, I'd prefer to do it sooner rather than later. (kinda afraid of forgetting about this too). If after we convert to promises, we are able to remove `trailing_metadata_available` field completely, then we could simply delete these six lines from call.cc